File: Python/Palindromes/Permutation_Palindrome.py
# Determine whether a given string's any permutation of itself is a palindrome or not
# author: Pavan Kumar Paluri
# Leetcode Question: https://leetcode.com/problems/palindrome-permutation/
# Using Hashmap : Time Complexity O(n), Space Complexity: O(1) - [Because, the number of distinct characters in the map as keys are limited in number]

from collections import Counter
class Solution:
    def canPermutePalindrome(self, s: str) -> bool:
        h_map = Counter(s)
        count = 0
        for val in h_map.values():
            count += val%2
        if count ==0 or count ==1:
            return True
        else:
            return False

File: C++/Palindromes/Permutation_Palindrome.cpp
// Permutation Palindrome 
// Author: Pavan Kumar Paluri
// Time Complexity: O(N) and Space: O(1) - [Only a hashmap of 26 characters is possible as there are only limited set of chars]
// Leetcode Question: https://leetcode.com/problems/palindrome-permutation/solution/

class Solution {
public:
    bool canPermutePalindrome(string s) {
        // Create an unordered map
        unordered_map<char,int>h_map;
        for(int i=0; i<s.size(); i++)
        {
            h_map[s[i]]++;
        }
        int counter = 0;
        for(auto iter:h_map)
        {
            counter += iter.second%2;
        }
        if(counter==0 || counter ==1)
            return true;
        else
            return false;
    }
};
